pkg/controller/uniteddeployment/allocator.go

```diff
@@ -302,58 +303,220 @@ func (ac *elasticAllocator) validateAndCalculateMinMaxMap(replicas int32, nameTo
+	for _, subset := range ac.Spec.Topology.Subsets {
+		subsets = append(subsets, subset.Name)
+	}
+	nextReplicas := allocateByMinMaxMapAndReservation(replicas, minReplicasMap, maxReplicasMap, existingSubsets, subsets)
+	klog.V(4).InfoS("got UnitedDeployment next replicas", "unitedDeployment",
+		klog.KObj(ac.UnitedDeployment), "nextReplicas", nextReplicas)
+	return nextReplicas, nil
+}
+
+type reservationAllocatorSubsetRecord struct {
+	maxAllocatable int32 // Refer to `allocateByMinMaxMapAndReservation` for it's usage
+	replicas       int32 // Allocated replicas of the subset, which will be set to the value of the workload's spec.replicas
+}
+
+// allocate allocates `to` replicas to a subset and returns how many replicas is really allocated.
+//
+// Under normal circumstances, if X replicas are allocated to a subset, the number of remaining unassigned replicas
+// should obviously decrease by X. However, if they are assigned to an unschedulable subset that only has Y healthy Pods,
+// which is stored in field maxAllocatable, then starting from the (Y+1)th replica, they will certainly be reserved Pods
+// that cannot be scheduled. Therefore, when X (X > Y) replicas are allocated to such an unschedulable subset, the number
+// of remaining unassigned replicas can only decrease by Y. This means that the X-Y reserved replicas, which are expected
+// to fail to start, will be redundantly assigned in other subsets to create temporary Pods as their substitutes.
+// As the reserved Pods gradually recover, Y will gradually increase, and the number of temporary Pods, X-Y, will gradually
+// decrease until all work is handed back to the recovered reserved Pods.
+//
+// Following are some examples (Assume there are 5 replicas remaining to be allocated in total):
+//
+//  1. Trying to allocate 3 replicas to a subset with 0 healthy Pods and 2 reserved Pods (typical unschedulable):
+//     allocate(3, 5), -> 0,
+//     the number of remaining unassigned replicas is decreased by 0, 3-0=3 temporary pods will be created.
+//
+//  2. Trying to allocate 3 replicas to a subset with 3 healthy Pods and no reserved Pods (schedulable):
+//     allocate(3, 5), -> 3,
+//     the number of remaining unassigned replicas is decreased by 3, 3-3=0 temporary pods will be created.
+//
+//  3. Trying to allocate 3 replicas to a subset with 1 healthy Pod and 1 reserved Pod:
+//     allocate(3, 5), -> 1,
+//     the number of remaining unassigned replicas is decreased by 1, 3-1=2 temporary pod will be created
+//
+//     - param to: will scale up the subset to the value
+//     - param rest: rest replicas can be allocated
+//     - return allocated: the number of replicas that should be decreased. if not decreased, temporary pods will be created.
+func (ar *reservationAllocatorSubsetRecord) allocate(to, rest int32) (allocated int32) {
+	toAdd := min(to-ar.replicas, rest)
+	ar.replicas += toAdd
+	ar.maxAllocatable -= toAdd
+	return max(0, min(0, ar.maxAllocatable)+toAdd)
```

Comment:
nested min and max operator bring a heavy cognitive burden, the code should be written to be understand not to be confused. consider change it to a if clause and add related comments to each condition, e.g. 

```
if ar. maxAllocatable > 0 {
   return toAdd
} else {
   // comment what the negative maxAllocatable means here 
   return ar.maxAllocatable + toAdd
}
```